### DISCUSSION TITLE
Welcome to librespot-python Discussions!

### DISCUSSION THREAD

**Discussion by kokarare1212** [kokarare1212/librespot-python#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by shauneccles** [comment]
Hi!

I'm Shaun and I'm one of the LedFx developers.

LedFx is an LED visualisation project that turns music into light. We've always battled with audio sourcing, and current workarounds are pretty hacky - I've got a Chromecast audio plugged into my computer and have made a group with my other audio in order to play music in areas away from my computer.

I'm excited to see this python implementation - I hope in the future we can spin up a Spotify connect endpoint automatically and have users stream directly to it!

Thanks for all your work, I'll hopefully be able to contribute!

Shaun

**Reply by kokarare1212** [reply]
Hello.

Thank you for your interest in this project.

This project is based on [this project](https://github.com/librespot-org/librespot-java), and I hope you find it useful.
We designed it to be compatible with the base project, mainly as described in the [ROADMAP.md](https://github.com/kokarare1212/librespot-python/blob/main/ROADMAP.md), but some features are not available due to circular import.
Due to various other problems, we have not been able to release it to Pypi, and it is currently an alpha version.

We hope you can contribute to this project!

kokarare1212

**Comment by AbhayAysola** [comment]
Hi, 

This is a cool project but I couldn't help but ask how you implemented support for free accounts as well. Because afaik librespot-java doesn't support free accounts so I was just curious how that works.

**Reply by kokarare1212** [reply]
Hello.

Thank you for your interest in this project.

librespot-java uses a login from Spotify Connect, so a paid account is required.
However, librespot-python is implemented using the same login method as the traditional Spotify app, so it can work with a free account.
So, in effect, librespot-java also allows you to log in with a free account.

**Reply by AbhayAysola** [reply]
Are there any docs for the login method? I'm trying to login with a free account on librespot (rust) and I looked at the code for login on both the rust repo and this one, and they both do the same things, however for the rust repo the server returns a PremiumAccountRequired error. I also tried to check the bytes sent to the server and they are the exact same, which is very odd.

**Reply by kokarare1212** [reply]
There is no documentation (I plan to make one:)).
But as far as I know, there are three ways to login: 1. login with username and password, 2. login with facebook, 3. login by connecting to the speaker from the official Spotify app. The standard application of librespot and librespot-java uses method 3, and librespot-python mainly uses method 1. (Method 3 is also implemented, but it is almost useless;(...)

Also, the code corresponding to librespot-python's 1 is
https://github.com/kokarare1212/librespot-python/blob/rewrite/librespot/core.py#L950-L1011
and the equivalent code for 3 is
https://github.com/kokarare1212/librespot-python/blob/rewrite/librespot/zeroconf.py#L110-L188

**Reply by AbhayAysola** [reply]
Thanks for the quick replies, I've actually gone through the code for the 1st method in both the repos. 
The weird behavior is that in the rust code the server responds with APLoginFailed when I try to login with a free account, whereas with the python code, the server returns APWelcome. 
This is weird because I actually checked the bytes that were being sent to the servers, and the only difference I found was in the last 4 bits - MAC so I can't really imagine what the difference between the two could be. 

Anyway, I'd like to hear your thoughts on what's going on with this. Thanks again for replying and I hope I'll be able to contribute someday :)

**Comment by kiritannplum** [comment]
hello! I have a question.
However, I am very new to both Python and Github, so there may be many mistakes.

I wanted to control Spotify on the python and play songs, and after some research I came across this project.
I immediately tried to start "examples/player/main.py", but I got an error "'bytes' object cannot be interpreted as an integer" on line 130.
I understand that it seems to be failing when trying to convert the variable 'byte', but I have no idea what to do from there.
I can attach the contents of the variable 'byte', but I can't decide whether it's okay to attach it by myself because it may contain some security-critical data.
I would appreciate it if you could give me some advice.
Please let me know if you need any information. I will prepare it immediately.
Here is a list of information that may be of use to you at this time.

spotify premium account
windows10
Python 3.10.7
Using Visual Studio Code

[error statement]
'bytes' object cannot be interpreted as an integer
  File "[omitted]\librespot-python-main\examples\player\main.py", line 130, in play
    ffplay.stdin.write(bytes([byte]))
  File "[omitted]\librespot-python-main\examples\player\main.py", line 44, in client
    play(track_id_str)
  File "[omitted]\librespot-python-main\examples\player\main.py", line 143, in main
    client()
  File "[omitted]\librespot-python-main\examples\player\main.py", line 153, in <module>
    main()

**Reply by kokarare1212** [reply]
I have updated this example because it was based on an old library.
https://github.com/kokarare1212/librespot-python/issues/155

**Reply by kiritannplum** [reply]
I am sorry for my late reply.
I applied the #155 update, but unfortunately I was unable to play it.
However, after trying various things, I could play it by rewriting line 130 from 
ffplay.stdin.write(bytes([byte]))
to 
ffplay.stdin.write(byte)

**Reply by kokarare1212** [reply]
Yes, the type returned by read is bytes, so I reused it.

**Reply by 0xthemiddleman** [reply]
I'm trying to generate proto class using this command:

protoc --include_imports --python_out=../librespot/proto *.proto

but the gained class are different from those of that repository, for example this is the auth.proto:

# -*- coding: utf-8 -*-
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# source: authentication.proto
"""Generated protocol buffer code."""
from google.protobuf.internal import builder as _builder
from google.protobuf import descriptor as _descriptor
from google.protobuf import descriptor_pool as _descriptor_pool
from google.protobuf import symbol_database as _symbol_database
# @@protoc_insertion_point(imports)

_sym_db = _symbol_database.Default()




DESCRIPTOR = _descriptor_pool.Default().AddSerializedFile(b'\n\x14\x61uthentication.proto\x12\x07spotify\"\xa2\x03\n\x17\x43lientResponseEncrypted\x12\x34\n\x11login_credentials\x18\n \x02(\x0b\x32\x19.spotify.LoginCredentials\x12\x32\n\x10\x61\x63\x63ount_creation\x18\x14 \x01(\x0e\x32\x18.spotify.AccountCreation\x12?\n\x14\x66ingerprint_response\x18\x1e \x01(\x0b\x32!.spotify.FingerprintResponseUnion\x12-\n\x0bpeer_ticket\x18( \x01(\x0b\x32\x18.spotify.PeerTicketUnion\x12(\n\x0bsystem_info\x18\x32 \x02(\x0b\x32\x13.spotify.SystemInfo\x12\x16\n\x0eplatform_model\x18< \x01(\t\x12\x16\n\x0eversion_string\x18\x46 \x01(\t\x12)\n\x06\x61ppkey\x18P \x01(\x0b\x32\x19.spotify.LibspotifyAppKey\x12(\n\x0b\x63lient_info\x18Z \x01(\x0b\x32\x13.spotify.ClientInfo\"a\n\x10LoginCredentials\x12\x10\n\x08username\x18\n \x01(\t\x12(\n\x03typ\x18\x14 \x02(\x0e\x32\x1b.spotify.AuthenticationType\x12\x11\n\tauth_data\x18\x1e \x01(\x0c\"\x89\x01\n\x18\x46ingerprintResponseUnion\x12\x30\n\x05grain\x18\n \x01(\x0b\x32!.spotify.FingerprintGrainResponse\x12;\n\x0bhmac_ripemd\x18\x14 \x01(\x0b\x32&.spotify.FingerprintHmacRipemdResponse\"1\n\x18\x46ingerprintGrainResponse\x12\x15\n\rencrypted_key\x18\n \x02(\x0c\"-\n\x1d\x46ingerprintHmacRipemdResponse\x12\x0c\n\x04hmac\x18\n \x02(\x0c\"o\n\x0fPeerTicketUnion\x12\x30\n\npublic_key\x18\n \x01(\x0b\x32\x1c.spotify.PeerTicketPublicKey\x12*\n\nold_ticket\x18\x14 \x01(\x0b\x32\x16.spotify.PeerTicketOld\")\n\x13PeerTicketPublicKey\x12\x12\n\npublic_key\x18\n \x02(\x0c\"C\n\rPeerTicketOld\x12\x13\n\x0bpeer_ticket\x18\n \x02(\x0c\x12\x1d\n\x15peer_ticket_signature\x18\x14 \x02(\x0c\"\x81\x02\n\nSystemInfo\x12&\n\ncpu_family\x18\n \x02(\x0e\x32\x12.spotify.CpuFamily\x12\x13\n\x0b\x63pu_subtype\x18\x14 \x01(\r\x12\x0f\n\x07\x63pu_ext\x18\x1e \x01(\r\x12\x1d\n\x05\x62rand\x18( \x01(\x0e\x32\x0e.spotify.Brand\x12\x13\n\x0b\x62rand_flags\x18\x32 \x01(\r\x12\x17\n\x02os\x18< \x02(\x0e\x32\x0b.spotify.Os\x12\x12\n\nos_version\x18\x46 \x01(\r\x12\x0e\n\x06os_ext\x18P \x01(\r\x12!\n\x19system_information_string\x18Z \x01(\t\x12\x11\n\tdevice_id\x18\x64 \x01(\t\"p\n\x10LibspotifyAppKey\x12\x0f\n\x07version\x18\x01 \x02(\r\x12\x0e\n\x06\x64\x65vkey\x18\x02 \x02(\x0c\x12\x11\n\tsignature\x18\x03 \x02(\x0c\x12\x11\n\tuseragent\x18\x04 \x02(\t\x12\x15\n\rcallback_hash\x18\x05 \x02(\x0c\"X\n\nClientInfo\x12\x0f\n\x07limited\x18\x01 \x01(\x08\x12\'\n\x02\x66\x62\x18\x02 \x01(\x0b\x32\x1b.spotify.ClientInfoFacebook\x12\x10\n\x08language\x18\x03 \x01(\t\"(\n\x12\x43lientInfoFacebook\x12\x12\n\nmachine_id\x18\x01 \x01(\t\"\xe9\x02\n\tAPWelcome\x12\x1a\n\x12\x63\x61nonical_username\x18\n \x02(\t\x12\x34\n\x16\x61\x63\x63ount_type_logged_in\x18\x14 \x02(\x0e\x32\x14.spotify.AccountType\x12\x38\n\x1a\x63redentials_type_logged_in\x18\x19 \x02(\x0e\x32\x14.spotify.AccountType\x12\x43\n\x1ereusable_auth_credentials_type\x18\x1e \x02(\x0e\x32\x1b.spotify.AuthenticationType\x12!\n\x19reusable_auth_credentials\x18( \x02(\x0c\x12\x12\n\nlfs_secret\x18\x32 \x01(\x0c\x12*\n\x0c\x61\x63\x63ount_info\x18< \x01(\x0b\x32\x14.spotify.AccountInfo\x12(\n\x02\x66\x62\x18\x46 \x01(\x0b\x32\x1c.spotify.AccountInfoFacebook\"k\n\x0b\x41\x63\x63ountInfo\x12,\n\x07spotify\x18\x01 \x01(\x0b\x32\x1b.spotify.AccountInfoSpotify\x12.\n\x08\x66\x61\x63\x65\x62ook\x18\x02 \x01(\x0b\x32\x1c.spotify.AccountInfoFacebook\"\x14\n\x12\x41\x63\x63ountInfoSpotify\"?\n\x13\x41\x63\x63ountInfoFacebook\x12\x14\n\x0c\x61\x63\x63\x65ss_token\x18\x01 \x01(\t\x12\x12\n\nmachine_id\x18\x02 \x01(\t*\xd6\x01\n\x12\x41uthenticationType\x12\x1c\n\x18\x41UTHENTICATION_USER_PASS\x10\x00\x12-\n)AUTHENTICATION_STORED_SPOTIFY_CREDENTIALS\x10\x01\x12.\n*AUTHENTICATION_STORED_FACEBOOK_CREDENTIALS\x10\x02\x12 \n\x1c\x41UTHENTICATION_SPOTIFY_TOKEN\x10\x03\x12!\n\x1d\x41UTHENTICATION_FACEBOOK_TOKEN\x10\x04*Y\n\x0f\x41\x63\x63ountCreation\x12\"\n\x1e\x41\x43\x43OUNT_CREATION_ALWAYS_PROMPT\x10\x01\x12\"\n\x1e\x41\x43\x43OUNT_CREATION_ALWAYS_CREATE\x10\x03*\x9d\x01\n\tCpuFamily\x12\x0f\n\x0b\x43PU_UNKNOWN\x10\x00\x12\x0b\n\x07\x43PU_X86\x10\x01\x12\x0e\n\nCPU_X86_64\x10\x02\x12\x0b\n\x07\x43PU_PPC\x10\x03\x12\x0e\n\nCPU_PPC_64\x10\x04\x12\x0b\n\x07\x43PU_ARM\x10\x05\x12\x0c\n\x08\x43PU_IA64\x10\x06\x12\n\n\x06\x43PU_SH\x10\x07\x12\x0c\n\x08\x43PU_MIPS\x10\x08\x12\x10\n\x0c\x43PU_BLACKFIN\x10\t*K\n\x05\x42rand\x12\x13\n\x0f\x42RAND_UNBRANDED\x10\x00\x12\r\n\tBRAND_INQ\x10\x01\x12\r\n\tBRAND_HTC\x10\x02\x12\x0f\n\x0b\x42RAND_NOKIA\x10\x03*\xd1\x02\n\x02Os\x12\x0e\n\nOS_UNKNOWN\x10\x00\x12\x0e\n\nOS_WINDOWS\x10\x01\x12\n\n\x06OS_OSX\x10\x02\x12\r\n\tOS_IPHONE\x10\x03\x12\n\n\x06OS_S60\x10\x04\x12\x0c\n\x08OS_LINUX\x10\x05\x12\x11\n\rOS_WINDOWS_CE\x10\x06\x12\x0e\n\nOS_ANDROID\x10\x07\x12\x0b\n\x07OS_PALM\x10\x08\x12\x0e\n\nOS_FREEBSD\x10\t\x12\x11\n\rOS_BLACKBERRY\x10\n\x12\x0c\n\x08OS_SONOS\x10\x0b\x12\x0f\n\x0bOS_LOGITECH\x10\x0c\x12\n\n\x06OS_WP7\x10\r\x12\x0c\n\x08OS_ONKYO\x10\x0e\x12\x0e\n\nOS_PHILIPS\x10\x0f\x12\t\n\x05OS_WD\x10\x10\x12\x0c\n\x08OS_VOLVO\x10\x11\x12\x0b\n\x07OS_TIVO\x10\x12\x12\x0b\n\x07OS_AWOX\x10\x13\x12\x0c\n\x08OS_MEEGO\x10\x14\x12\r\n\tOS_QNXNTO\x10\x15\x12\n\n\x06OS_BCO\x10\x16*(\n\x0b\x41\x63\x63ountType\x12\x0b\n\x07Spotify\x10\x00\x12\x0c\n\x08\x46\x61\x63\x65\x62ook\x10\x01\x42\r\n\x0b\x63om.spotify')

_builder.BuildMessageAndEnumDescriptors(DESCRIPTOR, globals())
_builder.BuildTopDescriptorsAndMessages(DESCRIPTOR, 'authentication_pb2', globals())
if _descriptor._USE_C_DESCRIPTORS == False:

  DESCRIPTOR._options = None
  DESCRIPTOR._serialized_options = b'\n\013com.spotify'
  _AUTHENTICATIONTYPE._serialized_start=2083
  _AUTHENTICATIONTYPE._serialized_end=2297
  _ACCOUNTCREATION._serialized_start=2299
  _ACCOUNTCREATION._serialized_end=2388
  _CPUFAMILY._serialized_start=2391
  _CPUFAMILY._serialized_end=2548
  _BRAND._serialized_start=2550
  _BRAND._serialized_end=2625
  _OS._serialized_start=2628
  _OS._serialized_end=2965
  _ACCOUNTTYPE._serialized_start=2967
  _ACCOUNTTYPE._serialized_end=3007
  _CLIENTRESPONSEENCRYPTED._serialized_start=34
  _CLIENTRESPONSEENCRYPTED._serialized_end=452
  _LOGINCREDENTIALS._serialized_start=454
  _LOGINCREDENTIALS._serialized_end=551
  _FINGERPRINTRESPONSEUNION._serialized_start=554
  _FINGERPRINTRESPONSEUNION._serialized_end=691
  _FINGERPRINTGRAINRESPONSE._serialized_start=693
  _FINGERPRINTGRAINRESPONSE._serialized_end=742
  _FINGERPRINTHMACRIPEMDRESPONSE._serialized_start=744
  _FINGERPRINTHMACRIPEMDRESPONSE._serialized_end=789
  _PEERTICKETUNION._serialized_start=791
  _PEERTICKETUNION._serialized_end=902
  _PEERTICKETPUBLICKEY._serialized_start=904
  _PEERTICKETPUBLICKEY._serialized_end=945
  _PEERTICKETOLD._serialized_start=947
  _PEERTICKETOLD._serialized_end=1014
  _SYSTEMINFO._serialized_start=1017
  _SYSTEMINFO._serialized_end=1274
  _LIBSPOTIFYAPPKEY._serialized_start=1276
  _LIBSPOTIFYAPPKEY._serialized_end=1388
  _CLIENTINFO._serialized_start=1390
  _CLIENTINFO._serialized_end=1478
  _CLIENTINFOFACEBOOK._serialized_start=1480
  _CLIENTINFOFACEBOOK._serialized_end=1520
  _APWELCOME._serialized_start=1523
  _APWELCOME._serialized_end=1884
  _ACCOUNTINFO._serialized_start=1886
  _ACCOUNTINFO._serialized_end=1993
  _ACCOUNTINFOSPOTIFY._serialized_start=1995
  _ACCOUNTINFOSPOTIFY._serialized_end=2015
  _ACCOUNTINFOFACEBOOK._serialized_start=2017
  _ACCOUNTINFOFACEBOOK._serialized_end=2080
# @@protoc_insertion_point(module_scope)
